### PR TITLE
Autocrypt Setup Message sending

### PR DESCRIPTION
**Pull request by Valodim**
This PR adds support for sending an Autocrypt Setup Message. I'm quite fond of the UI flow, although as always the wording might change here and there:

Sending an Autocrypt Setup Message

This is still missing some error handling. There are also a bunch of Kotlin techniques in here that are new for me, beware newbie mistakes :)

I used scene transitions in this layout, instead of ViewAnimators. It clutters up the preview more, but it's also more flexible with its transformations.

### CODE REVIEW

**Comment by Valodim** [comment]
The functionality here depends on autocrypt setup message support from https://github.com/open-keychain/open-keychain/pull/2304 and a corresponding new api call https://github.com/open-keychain/open-keychain/pull/2305

- [ ] test with Enigmail
- [ ] add error handling for outdated version of OpenKeychain
- [ ] improve error handling if message can't be sent

I think the error handling can be fairly minimal: The semantics are basically all-or-nothing, if anything goes off the happy path here, something is misconfigured or broken and we don't want to proceed.

**Comment by violoncelloch** [comment]
Maybe test with Delta Chat?

**Comment by hpk42** [comment]
Regarding the error cases, it might be interesting to produce enough error information so that it can be pasted to a github issue and then makes sense for development.  

great work, btw!

**Comment by cketti** [comment]
UX comments:
- The text says "To finish, open the message on your other device and enter the transfer code". But the button reads "show setup code".
- The message itself contains "To set up your new device for Autocrypt, please follow the instuctions[typo!] that should be presented by your new device". That is confusing if the message is opened on the "new device". Enigmail displays a custom message that contains Enigmail-specific setup instructions. I like that approach.
- The "Autocrypt Setup Message" screen can only be exited using the device's back button. Maybe add an "up" button to the action bar?
- Decrypting and importing the key to OpenKeychain involved quite a few clicks. Also, the crypto provider then still has to be configured in K-9 Mail. I hope we can combine these two things in the future and make it require less clicks.

File: k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptKeyTransferActivity.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    private val openPgpApiManager: OpenPgpApiManager by inject { mapOf("lifecycle" to lifecycle) }
+    private lateinit var presenter: AutocryptKeyTransferPresenter
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.crypto_key_transfer)
+
+        val account = intent.getStringExtra(EXTRA_ACCOUNT)
+
+        transfer_send_button.setOnClickListener { presenter.onClickTransferSend() }
```
[the remark this answers is not in the document]
Comment (Valodim) [reply k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptKeyTransferActivity.kt]: I kind of liked how it made it immediately obvious that this is from a different namespace than other variables. It's used quite differently from any other import. I wouldn't mind using camelCase though, although I've never seen that used in ids in any other app

**Comment by Valodim** [comment]
> The message itself contains "To set up your new device for Autocrypt, please follow the instuctions[typo!] that should be presented by your new device". That is confusing if the message is opened on the "new device". Enigmail displays a custom message that contains Enigmail-specific setup instructions. I like that approach.

That text is actually the one that enigmail puts into the mail, including the typo :) I just copied it as a placeholder.

> Decrypting and importing the key to OpenKeychain involved quite a few clicks. Also, the crypto provider then still has to be configured in K-9 Mail. I hope we can combine these two things in the future and make it require less clicks.

Yes, that workflow is only what worked naturally when I changed openkeychain's backup format. This definitely needs specialized support, I just don't have a very good picture in mind how the flow should look. If you're interested, I'd love to brainstorm some time

File: k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptSetupMessageLiveEvent.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    private fun loadAutocryptSetupMessage(resources: Resources, openPgpApi: OpenPgpApi, account: Account): AutocryptSetupMessage {
+        val keyIds = longArrayOf(account.openPgpKey)
+        val address = Address.parse(account.getIdentity(0).email)[0]
+
+        val intent = Intent(OpenPgpApi.ACTION_AUTOCRYPT_KEY_TRANSFER)
+        intent.putExtra(OpenPgpApi.EXTRA_KEY_IDS, keyIds)
+        val baos = ByteArrayOutputStream()
+        val result = openPgpApi.executeApi(intent, null as InputStream?, baos)
```
[the remark this answers is not in the document]
Comment (Valodim) [reply k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptSetupMessageLiveEvent.kt]: I agree :)

File: k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptSetupMessageLiveEvent.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        val keyData = baos.toByteArray()
+        val pi: PendingIntent = result.getParcelableExtra(OpenPgpApi.RESULT_INTENT)
+
+        val setupMessage = AutocryptTransferMessageUtil.createAutocryptTransferMessage(resources, keyData, address)
+
+        return AutocryptSetupMessage(setupMessage, pi)
+    }
+}
+
+data class AutocryptSetupMessage(val setupMessage: Message, val showTransferCodePi: PendingIntent)
```
[the remark this answers is not in the document]
Comment (Valodim) [reply k9mail/src/main/java/com/fsck/k9/ui/endtoend/AutocryptSetupMessageLiveEvent.kt]: Hmm, well, the event is that the setup message was generated. I agree it's not great, but I can't think of a good one

**Pull request merged**